src/intptrcast.rs

```diff
@@ -34,24 +40,36 @@ impl GlobalStateInner {
         GlobalStateInner {
             int_to_ptr_map: Vec::default(),
             base_addr: FxHashMap::default(),
+            exposed: FxHashSet::default(),
             next_base_addr: STACK_ADDR,
+            permissive_provenance: config.permissive_provenance,
             strict_provenance: config.strict_provenance,
         }
     }
 }
 
 impl<'mir, 'tcx> GlobalStateInner {
-    pub fn ptr_from_addr(addr: u64, ecx: &MiriEvalContext<'mir, 'tcx>) -> Pointer<Option<Tag>> {
-        trace!("Casting 0x{:x} to a pointer", addr);
+    // Returns the `AllocId` that corresponds to the specified addr,
+    // or `None` if the addr is out of bounds
+    fn alloc_id_from_addr(ecx: &MiriEvalContext<'mir, 'tcx>, addr: u64) -> Option<AllocId> {
         let global_state = ecx.machine.intptrcast.borrow();
 
-        if global_state.strict_provenance {
-            return Pointer::new(None, Size::from_bytes(addr));
+        if addr == 0 {
+            return None;
         }
 
         let pos = global_state.int_to_ptr_map.binary_search_by_key(&addr, |(addr, _)| *addr);
-        let alloc_id = match pos {
-            Ok(pos) => Some(global_state.int_to_ptr_map[pos].1),
+
+        match pos {
+            Ok(pos) => {
+                let (_, alloc_id) = global_state.int_to_ptr_map[pos];
+
+                if !global_state.permissive_provenance || global_state.exposed.contains(&alloc_id) {
```

Comment:
If `!global_state.permissive_provenance`, this function should never even be called.

Reply:
So this is actually very fun... This code passes permissive provenance checking.

```rust
let x: i32 = 3;
let x_ptr = &x as *const i32;

let x_usize = x_ptr as usize;

let ptr: *const i32 = unsafe { std::mem::transmute(x_usize) };
assert_eq!(unsafe { *ptr }, 3);
```

It appears as-if whatever does interpreting calls this function on every load of a pointer from a `Scalar`, which basically means an implicit int-to-ptr cast on every load. I'll probably have to handle that if we want to fix the permissive provenance semantics, but that likely breaks some current code in Miri, so we probably need some `enforce_xxx` flag for this.

Reply:
"This function" being `alloc_id_from_addr`. (GH makes this really hard to find out in the discussion view.)

This is related to what I said recently in the rustc PR: you need a separate hook on int2ptr casts. `ptr_from_addr` is then only called for ptr2int *transmutes* and should always return a `None` provenance when permissive provenance is enabled. (I don't think we want a separate flag. Code that does this *should* be broken.)

Reply:
The fun thing about that is function pointers. We don't allow as-casts between them and usize, so the test suite doesn't pass as it transmutes usizes to function pointers. I'm also not actually sure how to special case function pointers in `intptrcast.rs`, or if that is the right move.

Reply:
I thought we allow casting them to raw ptrs and those to usize, don't we?

Reply:
It's the usize -> fn direction that we don't allow.

```
error[[E0605]](https://doc.rust-lang.org/stable/error-index.html#E0605): non-primitive cast: `usize` as `fn()`
 --> src/main.rs:4:13
  |
4 |     let _ = 555usize as fn() -> ();
  |             ^^^^^^^^^^^^^^^^^^^^^^ invalid cast
```

Reply:
Hm, and `*const ()` as `fn()` also does not seem to work.
For now, the right thing is to first `from_exposed_addr` to `*const ()` and then transmute to `fn()`, but we clearly need better APIs here.

Reply:
(Or of course, the other alternative is to keep this at ptr type all along, and never get a `usize`. That is probably better.)

Reply:
Well, never using `usize` would be a great alternative in my book too, but that seems kind of difficult, especially since this is stuff breaking in tests, namely `ptr_offset.rs`  (and very likely in real world code).

I'm not sure why being able to cast from usize was never implemented for function pointers though... It also appears that fn ptrs currently have no methods, which is going to be fun if we want to ship any new APIs.